src/backend/vault-hsm/dh-newhope.c

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright (C) 2016 Southern Storm Software, Pty Ltd.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+#include "internal.h"
+#include "crypto/newhope/newhope.h"
+#include <string.h>
+
+#define MAX_OF(a, b) ((a) > (b) ? (a) : (b))
+
+typedef struct NoiseNewHopeState_s
+{
+    struct NoiseDHState_s parent;
+    uint8_t random_data[64];
+    uint16_t generated;
+    poly private_key;
+    uint8_t public_key[MAX_OF(NEWHOPE_SENDABYTES, NEWHOPE_SENDBBYTES)];
+
+} NoiseNewHopeState;
+
+static int noise_newhope_generate_keypair
+    (NoiseDHState *state, const NoiseDHState *other)
+{
+    NoiseNewHopeState *st = (NoiseNewHopeState *)state;
+    NoiseNewHopeState *os = (NoiseNewHopeState *)other;
+    if (st->parent.role == NOISE_ROLE_RESPONDER) {
+        /* Generating the keypair for Bob relative to Alice's parameters */
+        if (!os || os->parent.key_type == NOISE_KEY_TYPE_NO_KEY)
+            return NOISE_ERROR_INVALID_STATE;
+        noise_rand_bytes(st->random_data, st->parent.private_key_len);
```

Comment:
Why do you ensure that the `other` pointer is not `NULL` before de-referencing it (through the `if (!os||...)` statement :arrow_up:), but you don't check the validity of the `st` pointer before accessing it? Is that latter guaranteed by caller to be non-NULL?